Fix Swift dictionary padding with enum keys

keith opened the pull request:
When you have a dictionary which has an enum as the key such as:

```
enum Foo: String {
  case a
}

let bar = [Foo.a: 1]
```

The stride of the key is 1 byte. This differs from if you have a tuple
with the same key and value as the dictionary such as:

```
let baz = (Foo.a: 1)
```

In order to read the values from a dictionary we create a tuple which
expects 16 bytes of data (in the case the key is an enum and the value
is an int). Previously because the key was only 1 byte, we only filled
it with 9 bytes, and therefore the offsets were wrong leading to us
printing garbage data. Now instead of just using the stride of the key,
we use the stride based on the tuple type we generate for printing these
values. This works since the buffer is initialized to all 0 and
therefore acts as padding for the key.

This fixes https://bugs.swift.org/browse/SR-6290

keith committed:
Fix Swift dictionary padding with enum keys

When you have a dictionary which has an enum as the key such as:

```
enum Foo: String {
  case a
}

let bar = [Foo.a: 1]
```

The stride of the key is 1 byte. This differs from if you have a tuple
with the same key and value as the dictionary such as:

```
let baz = (Foo.a: 1)
```

In order to read the values from a dictionary we create a tuple which
expects 16 bytes of data (in the case the key is an enum and the value
is an int). Previously because the key was only 1 byte, we only filled
it with 9 bytes, and therefore the offsets were wrong leading to us
printing garbage data. Now instead of just using the stride of the key,
we use the stride based on the tuple type we generate for printing these
values. This works since the buffer is initialized to all 0 and
therefore acts as padding for the key.

keith commented:
To test this I've been using this file:

```swift
enum Enum {
    case a, b, c
}

let testA = [Enum.c: 1, Enum.b: 2]
let testB = ["a": 1, "b": 2]
let testC = (key: Enum.b, value: 2)
let testD = Set([Enum.c])
print(testA)
```

And this `lldbinit` for convenience:

```
br set -f foo.swift -l 9
br command add 1 -o "fr v testA testB testC testD"
r
```

lattner commented:
Great fix @keith!!

DougGregor commented:
@swift-ci please test

keith commented:
Looks like the linux failures here might be unrelated (although I'm not sure if this is the actual failure cause):

```
fatal error: error in backend: IO failure on output stream.
```

The macOS failure looks like it could be related so I'm trying to repro these locally.

keith commented:
The macOS tests seem to pass for me locally on my branch. It looks like based on the log this one might have failed because of a timeout:

```
[TestSwiftDictionaryNSObjectAnyObject.py FAILED] (TIMEOUT)
```

Can we run these again if that makes sense?

CodaFi commented:
I'll kick off the tests, but that failure looks concerning.  It seems the printer is straying pretty far outside the bounds of the object it should be reading. I would guess your local setup would have trouble reproducing this because it would appear nondeterministically.

@swift-ci please test

keith commented:
Anyone have thoughts on why this would fail on CI both runs, but not fail locally in my testing?

keith committed:
Add tests for SwiftHashedContainer with enum keys

This tests how hashed containers behave with enums as keys.

keith commented:
I've submitted a related PR adding tests for this case. Since they should be valuable regardless of this fix (minus the one that is disabled until this bug is fixed) I thought that made sense to be in a separate PR https://github.com/apple/swift-lldb/pull/262

keith commented:
Ah it looks like some of the tests that failed here were disabled in this commit https://github.com/apple/swift-lldb/commit/909ec3e70551d7c5eb40632bd5020ebb87e32f77

keith commented:
I've rebased on the stable branch which has these changes, so if someone could run the tests again here they should be green this time (assuming the linux OOM failures have been fixed as well)

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

keith commented:
Is this linux failure a known issue?

```
clang: error: unable to execute command: Killed
clang: error: linker command failed due to signal (use -v to see invocation)
```

This is similar to how it has failed on Linux all 3 test runs so far

dduan commented:
@swift-ci please test

keith commented:
Looks like linux failed again for the same reason 🤔

CodaFi commented:
Linux failure doesn’t look infrastructure related.  @jimingham any ideas?

dcci reviewed (changes requested):
Hi @Keith, is there any change of adding a test for this?
Also, I looked at the failure and I'm pretty confident it's unrelated. 

Can you please consider rebasing the following patch on ToT? At that point I'll spawn another build. Sorry for hanging you for so long, but it seems there are more people looking at Swift debug support in LLDB than there used to be until very recently :)

dcci commented:
Nvm, I just noticed you submitted another pull request for the test. My comment on rebasing still stands.

keith commented:
@dcci I've rebased on `stable` (let me know if that's not what you meant) and added the test from https://github.com/apple/swift-lldb/pull/262 here instead.

Let me know if there's anything else I can improve here!

dcci commented:
Let's test this again and cross finger. If this still fails I'll take a look myself.

dcci commented:
@swift-ci please test

dcci commented:
@swift-ci please test

keith commented:
Looks like the Linux build failed with the same error as before

dcci commented:
@Keith yes, definitely not related. On a bus now but I'll check in 30 where this problem is.

shahmishal commented:
@swift-ci test Linux

shahmishal commented:
Not sure, I have not seen this before. 

```
clang: error: unable to execute command: Killed
clang: error: linker command failed due to signal (use -v to see invocation)
```

dduan commented:
It certainly _looks_ like OOM. But why would this change cause that....

shahmishal commented:
@swift-ci test Linux

keith commented:
It passed 🎉  Thanks @shahmishal !

dcci commented:
@shahmishal thanks. Can we get the bot back online so I can merge this?

Thanks!

--
Davide

shahmishal commented:
You should be able to merge now.

dcci commented:
It was an hw issue, merging this now.

keith commented:
Awesome! Thanks everyone for your help!
